mesonbuild/wrap/wrap.py

```diff
@@ -97,25 +98,34 @@ def __init__(self, fname: str):
         self.provided_deps = {} # type: T.Dict[str, T.Optional[str]]
         self.provided_programs = [] # type: T.List[str]
         self.basename = os.path.basename(fname)
-        self.name = self.basename
-        if self.name.endswith('.wrap'):
-            self.name = self.name[:-5]
+        self.has_wrap = self.basename.endswith('.wrap')
+        self.name = self.basename[:-5] if self.has_wrap else self.basename
+        self.directory = self.name
         self.provided_deps[self.name] = None
-        if fname.endswith('.wrap'):
-            self.parse_wrap(fname)
+        if self.has_wrap:
+            self.parse_wrap()
         self.directory = self.values.get('directory', self.name)
         if os.path.dirname(self.directory):
             raise WrapException('Directory key must be a name and not a path')
         if self.type and self.type not in ALL_TYPES:
             raise WrapException('Unknown wrap type {!r}'.format(self.type))
+        self.filesdir = os.path.join(os.path.dirname(self.filename), 'packagefiles')
 
-    def parse_wrap(self, fname: str) -> None:
+    def parse_wrap(self) -> None:
         try:
             self.config = configparser.ConfigParser(interpolation=None)
-            self.config.read(fname)
+            self.config.read(self.filename)
         except configparser.Error:
             raise WrapException('Failed to parse {}'.format(self.basename))
         self.parse_wrap_section()
+        if self.type == 'redirect':
+            # [wrap-redirect] have a `filename` value pointing to the real wrap
+            # file we should parse instead. It can be either absolute path or
+            # relative to the current wrap file location.
+            dirname = os.path.dirname(self.filename)
+            self.filename = os.path.join(dirname, self.values['filename'])
```

Comment:
This should validate that the value points to an actual subproject and does not try to do anything nasty like having `..` in it. Because if we don't check, people will abuse the crap out of it. The path should thus be of the type `subprojects/foo/subprojects/bar` or it should have only the project names like `foo/bar` or, if we had the syntax from #7513,  `['foo', 'bar']`.

There should also be a test for the forwarding wrap file by itself where the dependency info is static and in Git.

Reply:
Added unit tests and a few sanity checks:  filename must be within current subprojects/ dir, must point to a .wrap file that exists.

I'm against the form `foo/bar` because it looks like a path but it's not, that's too confusing. Took me a while to understand your suggestion. `['foo', 'bar']` is a bit better but not currently possible. `subprojects/foo/subprojects/bar` it should be in that form but it's actually hard to enforce because each subproject could have subprojects_dir set to a different value. Those redirect wrap files are parsed at the very beginning of the main project interpreter, at that point it has no idea of any subprojects.

Reply:
> it should be in that form but it's actually hard to enforce because each subproject could have subprojects_dir set to a different value

Let me tell you a little secret. There is at least one other place in the code that does this kind of traversal and it completely ignores the fact that `subprojects_dir` can be changed. Yet, no-one has ever complained or filed a bug. The only people who use that seem to be existing corporate monorepo thingies that will never be used as subprojects themselves. In fact, we could even deprecate that setting, it is only meant as a transitory helper.

Summa summarum: the actual syntax is not that important, but it really must validate that nested subprojects come from the respective subprojects dir rather than some random dir elsewhere.

Reply:
Ok then... updated the PR to enforce it.